File: handy-python/quiz/q17_number_of_islands_medium.py
"""
Q17) Number of Islands

AI-BEST:
- Scan the grid and flood-fill each new island with DFS/BFS.
- Time: O(rows * cols), Space: O(rows * cols) worst case

AI-EASY:
- Assign island labels manually and merge equivalent labels with union-find.
- Time: about O(rows * cols * alpha(n)), Space: O(rows * cols)
"""


def number_of_islands_best(grid: list[list[str]]) -> int:
    if not grid or not grid[0]:
        return 0

    rows = len(grid)
    cols = len(grid[0])
    islands = 0

    # Step 1: Scan every cell in the grid.
    # Why: We need to discover each island exactly once.
    for r in range(rows):
        for c in range(cols):
            if grid[r][c] != "1":
                continue

            # Step 2: When land appears, flood-fill it and mark the whole island visited.
            # Why: This prevents counting the same island multiple times.
            islands += 1
            stack = [(r, c)]
            grid[r][c] = "0"
            while stack:
                row, col = stack.pop()
                for dr, dc in ((1, 0), (-1, 0), (0, 1), (0, -1)):
                    nr = row + dr
                    nc = col + dc
                    if 0 <= nr < rows and 0 <= nc < cols and grid[nr][nc] == "1":
                        grid[nr][nc] = "0"
                        stack.append((nr, nc))

    # Step 3: Return the total count after the full scan.
    # Why: Every land mass has now been visited and counted once.
    return islands


def number_of_islands_easy(grid: list[list[str]]) -> int:
    if not grid or not grid[0]:
        return 0

    rows = len(grid)
    cols = len(grid[0])
    labels = [[0] * cols for _ in range(rows)]
    parent: dict[int, int] = {}
    next_label = 1

    def find(label: int) -> int:
        while parent[label] != label:
            parent[label] = parent[parent[label]]
            label = parent[label]
        return label

    def union(a: int, b: int) -> None:
        root_a = find(a)
        root_b = find(b)
        if root_a != root_b:
            parent[root_b] = root_a

    # Step 1: Assign a label to each land cell, looking only at already-seen neighbors.
    # Why: A row-major scan can reuse labels from the top and left without extra traversal logic.
    for r in range(rows):
        for c in range(cols):
            if grid[r][c] != "1":
                continue

            neighbors = []
            if r > 0 and labels[r - 1][c] > 0:
                neighbors.append(labels[r - 1][c])
            if c > 0 and labels[r][c - 1] > 0:
                neighbors.append(labels[r][c - 1])

            if not neighbors:
                labels[r][c] = next_label
                parent[next_label] = next_label
                next_label += 1
            else:
                labels[r][c] = min(neighbors)
                for label in neighbors:
                    union(labels[r][c], label)

    # Step 2: Merge equivalent labels with union-find.
    # Why: Two scan paths can describe the same island, so we need a cheap way to collapse them.
    roots: set[int] = set()
    for r in range(rows):
        for c in range(cols):
            if labels[r][c] > 0:
                roots.add(find(labels[r][c]))

    # Step 3: Count the unique final roots.
    # Why: Each root represents one island after all merges are resolved.
    return len(roots)


def run_demo() -> None:
    print("Q17: Number of Islands")
    samples = [
        [
            ["1", "1", "0", "0", "0"],
            ["1", "1", "0", "0", "0"],
            ["0", "0", "1", "0", "0"],
            ["0", "0", "0", "1", "1"],
        ],
        [["1", "0", "1", "0"], ["0", "1", "0", "1"]],
        [["0", "0"], ["0", "0"]],
    ]
    for grid in samples:
        print(f"Input: grid={grid}")
        best_grid = [row[:] for row in grid]
        easy_grid = [row[:] for row in grid]
        print("  BEST:", number_of_islands_best(best_grid))
        print("  EASY:", number_of_islands_easy(easy_grid))


if __name__ == "__main__":
    run_demo()

File: handy-ts/quiz/q17_number_of_islands_medium.ts
/**
 * Q17) Number of Islands
 *
 * AI-BEST:
 * Flood fill from each unvisited land cell.
 * Time: O(rows * cols), Space: O(rows * cols) worst case
 *
 * AI-EASY:
 * Assign provisional island IDs and merge touching land with union-find.
 * Time: O(rows * cols * alpha(n)), Space: O(rows * cols)
 */

function numIslandsBest(grid: string[][]): number {
  if (grid.length === 0 || grid[0].length === 0) return 0;

  const rows = grid.length;
  const cols = grid[0].length;
  let islands = 0;

  function floodFill(row: number, col: number): void {
    if (row < 0 || row >= rows || col < 0 || col >= cols) return;
    if (grid[row][col] !== "1") return;

    grid[row][col] = "0";
    floodFill(row + 1, col);
    floodFill(row - 1, col);
    floodFill(row, col + 1);
    floodFill(row, col - 1);
  }

  // Step 1: Scan every cell because any land cell could begin a new island.
  for (let row = 0; row < rows; row++) {
    for (let col = 0; col < cols; col++) {
      if (grid[row][col] !== "1") continue;

      // Step 2: Count the island and flood fill it because we must mark the whole connected component as visited.
      islands++;
      floodFill(row, col);
    }
  }

  // Step 3: Return the total because every island has now been counted exactly once.
  return islands;
}

function numIslandsEasy(grid: string[][]): number {
  if (grid.length === 0 || grid[0].length === 0) return 0;

  const rows = grid.length;
  const cols = grid[0].length;
  const totalCells = rows * cols;
  const uf = new UnionFind(totalCells);
  const land = new Array<boolean>(totalCells).fill(false);

  // Step 1: Assign a provisional ID to each land cell because the easy approach needs explicit bookkeeping before merging.
  for (let row = 0; row < rows; row++) {
    for (let col = 0; col < cols; col++) {
      if (grid[row][col] === "1") {
        const id = row * cols + col;
        land[id] = true;
        uf.activate(id);
      }
    }
  }

  // Step 2: Merge touching land cells because neighboring pieces can belong to the same island.
  for (let row = 0; row < rows; row++) {
    for (let col = 0; col < cols; col++) {
      if (grid[row][col] !== "1") continue;
      const id = row * cols + col;
      if (row + 1 < rows && grid[row + 1][col] === "1") uf.union(id, (row + 1) * cols + col);
      if (col + 1 < cols && grid[row][col + 1] === "1") uf.union(id, row * cols + (col + 1));
    }
  }

  const roots = new Set<number>();
  for (let id = 0; id < totalCells; id++) {
    if (!land[id]) continue;
    roots.add(uf.find(id));
  }

  // Step 3: Count unique roots because each root represents one final island label.
  return roots.size;
}

class UnionFind {
  private readonly parent: number[];
  private readonly rank: number[];

  constructor(size: number) {
    this.parent = new Array<number>(size).fill(-1);
    this.rank = new Array<number>(size).fill(0);
  }

  activate(x: number): void {
    if (this.parent[x] === -1) this.parent[x] = x;
  }

  find(x: number): number {
    if (this.parent[x] === x) return x;
    if (this.parent[x] === -1) return -1;
    this.parent[x] = this.find(this.parent[x]);
    return this.parent[x];
  }

  union(a: number, b: number): void {
    this.activate(a);
    this.activate(b);

    let rootA = this.find(a);
    let rootB = this.find(b);
    if (rootA === -1 || rootB === -1 || rootA === rootB) return;

    if (this.rank[rootA] < this.rank[rootB]) {
      [rootA, rootB] = [rootB, rootA];
    }

    this.parent[rootB] = rootA;
    if (this.rank[rootA] === this.rank[rootB]) {
      this.rank[rootA]++;
    }
  }
}

function runDemo(): void {
  console.log("Q17: Number of Islands");
  const cases = [
    [
      ["1", "1", "0", "0", "0"],
      ["1", "1", "0", "0", "0"],
      ["0", "0", "1", "0", "0"],
      ["0", "0", "0", "1", "1"],
    ],
    [
      ["1", "1"],
      ["1", "1"],
    ],
    [
      ["0", "0", "0"],
    ],
  ];

  for (const grid of cases) {
    console.log("Input grid:", JSON.stringify(grid));
    const bestGrid = grid.map((row) => [...row]);
    const easyGrid = grid.map((row) => [...row]);
    console.log("  BEST:", numIslandsBest(bestGrid));
    console.log("  EASY:", numIslandsEasy(easyGrid));
  }
}

runDemo();

export {};
